fc_util/src/arg_parser.rs

```diff
@@ -0,0 +1,542 @@
+    }
+
+    // Populates the fields of ArgParser.
+    pub fn get_matches(mut self, mut args: Vec<String>) -> Result<Self> {
+        if args.contains(&HELP_ARG.to_string()) {
+            println!("Launch a microvm.");
+            for arg in &self.args_matches {
+                println!("{}", arg.format_help());
+            }
+            process::exit(0)
+        }
+
+        if let Some(index) = args.iter().position(|arg| arg.as_str() == DOUBLE_DASH) {
+            self.extra_args = args.split_at(index + 1).1.to_vec();
+            args = args.split_at(index).0.to_vec();
+        }
+
+        let mut index = 0;
+        while index < args.len() {
+            let arg = &args[index];
+            if !arg.starts_with(DOUBLE_DASH) {
+                return Err(Error::UnexpectedArgument(arg.clone()));
+            }
+            let arg_name = &arg[DOUBLE_DASH.len()..];
+
+            // Check if the argument is an expected one.
+            let arg_info = self
+                .args_matches
+                .get(arg_name)
+                .ok_or_else(|| Error::UnexpectedArgument(arg_name.to_string()))?;
+
+            // Check if the argument was not provided more than once.
+            if self
+                .args_matches
+                .get(arg_name)
+                .unwrap()
+                .user_value
+                .is_some()
+            {
+                return Err(Error::DuplicatedArgument(arg_name.to_string()));
+            }
+
+            let arg_value = if arg_info.takes_value {
+                // Check if user provided a value for the arguments that take a value.
+                if index + 1 >= args.len() || args[index + 1].starts_with(DOUBLE_DASH) {
+                    return Err(Error::MissingValue(arg_name.to_string()));
+                }
+                &args[index + 1]
+            } else {
+                FLAG_PROVIDED
+            };
+            let updated_arg = arg_info.clone().set_user_value(arg_value.to_string());
+
+            let step = if arg_info.takes_value { 2 } else { 1 };
+            index += step;
+
+            self.args_matches.replace(updated_arg);
+        }
+
+        for arg_info in self.args_matches.iter() {
+            // The arguments that are marked `required` must be provided by user.
+            if arg_info.required && arg_info.user_value.is_none() {
+                return Err(Error::MissingArgument(arg_info.name.to_string()));
+            }
+            // For the arguments that require a specific argument to be also present in the list
+            // of arguments provided by user, search for that argument.
+            if arg_info.user_value.is_some() {
+                if let Some(arg_name) = arg_info.requires {
+                    if !args.contains(&(format!("--{}", arg_name))) {
+                        return Err(Error::MissingArgument(arg_name.to_string()));
+                    }
+                }
+            }
+        }
+        Ok(self)
+    }
```

Comment:
There's a lot of logic here. It would be nice to split it in a couple of smaller methods if possible.